### PR TITLE
UCT/BASE/CUDA/GTEST: Introduce CUDA-IPC keepalive

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Introduce CUDA-IPC keepalive.

## Why ?

To use CUDA-IPC transport in UCP EPs created with error-handling support.

## How ?

1. Moved KA-related code implemented in SM and MM to the common code in `uct/base/uct_iface.[ch]`.
2. Split MM ep_check to two phases `create` and `ep_check` itself.
3. Re-used the common code in MM transports (i.e. their common code).
4. Re-used the common code in CUDA-IPC transport:
- create keepalive in `uct_ep_create()`, because there is no access them to pid/startime of a peer (this data is packed as UCT iface address)
- use common ep_check functionality in `uct_ep_check()`
5. Update CUDA-IPC iface address to include process start time (`ucs_time_t`) in addition to self process ID (`pid_t`).
6. Updated UCT keepalive gtest to test CUDA-IPC + adoption.
7. Updated `test_obj_size.cc` to check CUDA IPC EP object size.
8. Removed `uct_cuda_ipc_ep_addr_t` structure declaration.

### CODE REVIEW

**Comment by dmitrygx** [comment]
Azure CI gpu failure is unrelated to this PR - #6357

**Comment by dmitrygx** [comment]
@bureddy @Akshay-Venkatesh could you review pls?

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
bot:pipe:retest

File: src/uct/base/uct_iface.c [hunk 1]
```diff
@@ -586,3 +586,58 @@ ucs_status_t uct_base_ep_am_short_iov(uct_ep_h ep, uint8_t id, const uct_iov_t *
...
+    ucs_time_t createtime;
+
+    UCT_EP_KEEPALIVE_CHECK_PARAM(flags, comp);
+
+    ucs_assert(ka != NULL);
+
+    status = ucs_sys_get_file_time(ka->proc, UCS_SYS_FILE_TIME_CTIME,
+                                   &createtime);
+    if ((status != UCS_OK) || (ka->starttime != createtime)) {
+        status = UCS_ERR_ENDPOINT_TIMEOUT;
```
Comment (brminich) [inline src/uct/base/uct_iface.c:635]: return `uct_iface_handle_ep_err(tl_ep->iface, tl_ep, UCS_ERR_ENDPOINT_TIMEOUT)` would be shorter

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.h [hunk 2]
```diff
@@ -14,6 +14,7 @@
 
 #include "cuda_ipc_md.h"
 #include "cuda_ipc_ep.h"
+#include "cuda_ipc_cache.h"
```
Comment (brminich) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.h:17]: why is this needed?

File: src/uct/base/uct_iface.c (hunk 1 above)
Comment (dmitrygx) [reply src/uct/base/uct_iface.c:635]: fixed

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.h (hunk 2 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_ipc/cuda_ipc_iface.h:17]: it was done to not include cuda headers under `extern "C" {}` in test, because they use some templates inside.

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 3]
```diff
@@ -22,19 +22,30 @@
...
     uct_cuda_ipc_iface_t *iface = ucs_derived_of(params->iface,
                                                  uct_cuda_ipc_iface_t);
+    const uct_cuda_ipc_iface_addr_t *iface_addr;
 
     UCT_EP_PARAMS_CHECK_DEV_IFACE_ADDRS(params);
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
 
+    iface_addr = (const uct_cuda_ipc_iface_addr_t*)params->iface_addr;
+
+    self->keepalive = uct_ep_keepalive_create(iface_addr->pid,
```
Comment (Akshay-Venkatesh) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:37]: Is `uct_cuda_ipc_iface_get_address` guaranteed to be called before this point? What I see in UCP is worker_iface_open calling uct_iface_open and there isn't a call for uct_iface_get_address call before that. Isn't it better to just populate pid and starttime at iface_open call and make `uct_cuda_ipc_iface_get_address` a no-op?

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c [hunk 4]
```diff
@@ -68,18 +68,34 @@ ucs_status_t uct_cuda_ipc_iface_get_device_address(uct_iface_t *tl_iface,
...
-    *(pid_t*)iface_addr = getpid();
+    uct_cuda_ipc_iface_addr_t *addr = (uct_cuda_ipc_iface_addr_t*)iface_addr;
+    ucs_status_t status;
+    char proc[32];
+
+    addr->pid = getpid();
+
+    uct_ep_get_process_proc_dir(proc, sizeof(proc), addr->pid);
+    status = ucs_sys_get_file_time(proc, UCS_SYS_FILE_TIME_CTIME,
+                                   &addr->starttime);
```
Comment (Akshay-Venkatesh) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.c:79]: Looks like we're using starttime here to populate keepalive_info_t. Why not populate starttime as part of keepalive_create here https://github.com/openucx/ucx/pull/6352/files#diff-6e9feaecde4caffbd288eaf02c7ee17b6d183cbcfa0fcfaf2dff256332387fdfR614? We seem to using ka.starttime for all purposes and not really using cuda_iface_addr_t->starttime for anything other than populating ka.starttime

**Review by Akshay-Venkatesh — COMMENTED** [review]
@dmitrygx 

For cuda* transports, is it possible to hold a reference to the current device context at the point of keepalive_create call and check if the current context matches the cached context as part of keepalive_check? 

Also, what about keepalive support in cuda_copy and gdr_copy UCTs? do we not need similar additions?

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 3 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:37]: > Is uct_cuda_ipc_iface_get_address guaranteed to be called before this point?

yes, it's guaranteed. this iface_address received from a peer (via out of band communication or WIREUP_MSG phase in UCP).
so, it means that all fields should be initialized by our peer.

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c (hunk 4 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_ipc/cuda_ipc_iface.c:79]: iface_address is available only during `uct_ep_create()`, then we don't have an access to it
I preferred in the current implementation not to save pid/starttime to an additional structure to initialize keepalive in `uct_ep_check()` as we do in UCT/MM transport, because I don't see memory consumption improvements.

**Comment by dmitrygx** [comment]
> For cuda* transports, is it possible to hold a reference to the current device context at the point of keepalive_create call and check if the current context matches the cached context as part of keepalive_check (in addition to checking process createtime)?

so, does it mean that if a device context changed - some CUDA-IPC internal things will be broken?
I think if the context changed, it should still work right? Although if CUDA-IPC transport relies on this, it sounds like a good idea and we could check it.

> Also, what about keepalive support in cuda_copy and gdr_copy UCTs? do we not need similar additions?

they don't need, because they are used inside the same process. UCP don't check keepalive capabilities for them and don't require error handling support.

**Comment by dmitrygx** [comment]
> > For cuda* transports, is it possible to hold a reference to the current device context at the point of keepalive_create call and check if the current context matches the cached context as part of keepalive_check (in addition to checking process createtime)?
> 
> so, does it mean that if a device context changed - some CUDA-IPC internal things will be broken?
> I think if the context changed, it should still work right? Although if CUDA-IPC transport relies on this, it sounds like a good idea and we could check it.

@Akshay-Venkatesh correct me if I'm wrong, it seems that CUDA-IPC has the hash of caches of peer-access regions and those caches are per `pid:device` pairs. Also, each UCT EP should belong to the same CUDA device context during its "life". So, the idea is to cache `CUdevice` inside CUDA-IPC's EP and check that `cuCtxGetDevice()` returns the same context every `uct_ep_check()` call. Is it correct?

**Comment by Akshay-Venkatesh** [comment]
>  so, does it mean that if a device context changed - some CUDA-IPC internal things will be broken?
>  I think if the context changed, it should still work right?

We allow for a single thread to manage at most a single unique device context. A process can manage multiple devices if multiple threads manage unique device contexts and each thread belongs to separate UCP context. A single thread cannot use multiple devices or multiple device contexts. No two threads can have a device context for the same device. With these constraints, it is unsupported for a thread in a UCX process to change its device context.

 > Although if CUDA-IPC transport relies on this, it sounds like a good idea and we could check it.

If I understand keepalive protocol correctly, we may need to check this to see if the context that the current thread is using is the same as what was cached at starttime. A possible scenario is that during the application run, a device failure occurred and the application handled this and created a new context. If the UCT endpoint detects a new context, it should purge all of its previous cache state and create new mappings. If not, we would likely see memcpy errors. Is it keepalive feature's responsibility to purge and recover or just report error?

> correct me if I'm wrong, it seems that CUDA-IPC has the hash of caches of peer-access regions and those caches are per pid:device pairs. 

Correct. As per cuIpcOpenMemHandle documentation - "CUipcMemHandles from each CUdevice in a given process may only be opened by one CUcontext per CUdevice per other process". As we force a single thread to be bound to one device context, cache per (pid, device) pair suffices.

> Also, each UCT EP should belong to the same CUDA device context during its "life". So, the idea is to cache CUdevice inside CUDA-IPC's EP and check that cuCtxGetDevice() returns the same context every uct_ep_check() call. Is it correct?

Seems right. When multiple threads decide to use cuda_ipc endpoints, they should have either have the same device context or they should be threads belonging to different ucp_contexts and maybe handling different cuda device contexts but in that case the endpoint would be different. In both cases, there seems to be 1:1 mapping between device context and endpoint.

**Comment by dmitrygx** [comment]
> Is it keepalive feature's responsibility to purge and recover or just report error?

no, Keepalive can't handle it, it just reports an error to a user. So, UCP can take care of it.

> they should have either have the same device context or they should be threads belonging to different ucp_contexts and maybe handling different cuda device contexts but in that case the endpoint would be different. In both cases, there seems to be 1:1 mapping between device context and endpoint.

good, i tried to get a `CUdevice` inside `uct_ep_create()` through `UCT_CUDA_IPC_GET_DEVICE` macro which calls `cuCtxGetDevice()`, but it fails in our gtests. Should I do some prerequisites to initialize/set device/context?

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 3 above)
Comment (Akshay-Venkatesh) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:37]: > device

This should be handled by https://github.com/openucx/ucx/pull/5737 here https://github.com/openucx/ucx/pull/5737/files#diff-8b80dd5e092bb2439236b2f818c36fa956a3005025f6029c1ea9ce449d1ddf25R97

**Comment by Akshay-Venkatesh** [comment]
> > they should have either have the same device context or they should be threads belonging to different ucp_contexts and maybe handling different cuda device contexts but in that case the endpoint would be different. In both cases, there seems to be 1:1 mapping between device context and endpoint.
> 
> good, i tried to get a `CUdevice` inside `uct_ep_create()` through `UCT_CUDA_IPC_GET_DEVICE` macro which calls `cuCtxGetDevice()`, but it fails in our gtests. Should I do some prerequisites to initialize/set device/context?

     

This should be handled by #5737 here https://github.com/openucx/ucx/pull/5737/files#diff-8b80dd5e092bb2439236b2f818c36fa956a3005025f6029c1ea9ce449d1ddf25R97

**Comment by dmitrygx** [comment]
> > > they should have either have the same device context or they should be threads belonging to different ucp_contexts and maybe handling different cuda device contexts but in that case the endpoint would be different. In both cases, there seems to be 1:1 mapping between device context and endpoint.
> > 
> > 
> > good, i tried to get a `CUdevice` inside `uct_ep_create()` through `UCT_CUDA_IPC_GET_DEVICE` macro which calls `cuCtxGetDevice()`, but it fails in our gtests. Should I do some prerequisites to initialize/set device/context?
> 
> This should be handled by #5737 here https://github.com/openucx/ucx/pull/5737/files#diff-8b80dd5e092bb2439236b2f818c36fa956a3005025f6029c1ea9ce449d1ddf25R97

thank you!
I see, then I'll debug new code and understand why it fails.

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> > > > they should have either have the same device context or they should be threads belonging to different ucp_contexts and maybe handling different cuda device contexts but in that case the endpoint would be different. In both cases, there seems to be 1:1 mapping between device context and endpoint.
> > > 
> > > 
> > > good, i tried to get a `CUdevice` inside `uct_ep_create()` through `UCT_CUDA_IPC_GET_DEVICE` macro which calls `cuCtxGetDevice()`, but it fails in our gtests. Should I do some prerequisites to initialize/set device/context?
> > 
> > 
> > This should be handled by #5737 here https://github.com/openucx/ucx/pull/5737/files#diff-8b80dd5e092bb2439236b2f818c36fa956a3005025f6029c1ea9ce449d1ddf25R97
> 
> thank you!
> I see, then I'll debug new code and understand why it fails.

#5737 hasn't been merged yet.

**Comment by dmitrygx** [comment]
> > > > > they should have either have the same device context or they should be threads belonging to different ucp_contexts and maybe handling different cuda device contexts but in that case the endpoint would be different. In both cases, there seems to be 1:1 mapping between device context and endpoint.
> > > > 
> > > > 
> > > > good, i tried to get a `CUdevice` inside `uct_ep_create()` through `UCT_CUDA_IPC_GET_DEVICE` macro which calls `cuCtxGetDevice()`, but it fails in our gtests. Should I do some prerequisites to initialize/set device/context?
> > > 
> > > 
> > > This should be handled by #5737 here https://github.com/openucx/ucx/pull/5737/files#diff-8b80dd5e092bb2439236b2f818c36fa956a3005025f6029c1ea9ce449d1ddf25R97
> > 
> > 
> > thank you!
> > I see, then I'll debug new code and understand why it fails.
> 
> ~#5737 hasn't been merged yet.~ It was merged by the time I sent this message.

great :) will rebase on top of it
thank you @Akshay-Venkatesh

**Comment by dmitrygx** [comment]
> > > > > > they should have either have the same device context or they should be threads belonging to different ucp_contexts and maybe handling different cuda device contexts but in that case the endpoint would be different. In both cases, there seems to be 1:1 mapping between device context and endpoint.
> > > > > 
> > > > > 
> > > > > good, i tried to get a `CUdevice` inside `uct_ep_create()` through `UCT_CUDA_IPC_GET_DEVICE` macro which calls `cuCtxGetDevice()`, but it fails in our gtests. Should I do some prerequisites to initialize/set device/context?
> > > > 
> > > > 
> > > > This should be handled by #5737 here https://github.com/openucx/ucx/pull/5737/files#diff-8b80dd5e092bb2439236b2f818c36fa956a3005025f6029c1ea9ce449d1ddf25R97
> > > 
> > > 
> > > thank you!
> > > I see, then I'll debug new code and understand why it fails.
> > 
> > 
> > ~#5737 hasn't been merged yet.~ It was merged by the time I sent this message.
> 
> great :) will rebase on top of it
> thank you @Akshay-Venkatesh

@Akshay-Venkatesh it works fine
done, the implementation was changed as you suggested

**Comment by brminich** [comment]
failure seems to be relevant
```
2021-02-19T21:38:19.9443986Z [0xce95e000] local address length: 620
2021-02-19T21:38:19.9446046Z [1613770699.929170] [swx-rdmz-ucx-gpu-02:29294:0]     cuda_ipc_ep.c:37   UCX  ERROR   cuCtxGetDevice(&self->keepalive.device)() failed: invalid device context
2021-02-19T21:38:19.9447423Z Failed to ucp_ep_create
2021-02-19T21:38:19.9448128Z 
2021-02-19T21:38:20.0317933Z [1613770699.931110] [swx-rdmz-ucx-gpu-02:29294:0]      ucp_worker.c:2350 UCX  WARN  not all pending operations (1) were flushed on worker 0x2535640 that is being destroyed
2021-02-19T21:38:20.0319702Z [1613770699.956909] [swx-rdmz-ucx-gpu-02:29294:0]        rc_iface.c:713  UCX  WARN  some eps were not destroyed
2021-02-19T21:38:20.0320973Z [1613770699.957285] [swx-rdmz-ucx-gpu-02:29294:0]  rc_mlx5_common.c:581  UCX  WARN  mlx5dv_devx_obj_destroy(SRQ) failed: Invalid argument
2021-02-19T21:38:20.0322736Z [1613770699.998328] [swx-rdmz-ucx-gpu-02:29294:0]           mpool.c:44   UCX  WARN  object 0x2573d40 was not returned to mpool devx dbrec
2021-02-19T21:38:20.0323966Z [1613770700.013643] [swx-rdmz-ucx-gpu-02:29294:0]       ib_device.c:685  UCX  WARN  async_events_hash not empty
2021-02-19T21:38:30.0416957Z The STDIO streams did not close within 10 seconds of the exit event from process '/usr/bin/bash'. This may indicate a child process inherited the STDIO streams and has not yet exited.
2021-02-19T21:38:30.0480522Z ##[error]Bash exited with code '255'.
2021-02-19T21:38:35.3732189Z ##[section]Finishing: Run ./contrib/test_jenkins.sh
```

**Comment by dmitrygx** [comment]
@Akshay-Venkatesh @bureddy could you review pls?

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 5]
```diff
@@ -22,19 +22,47 @@
...
+    const uct_cuda_ipc_iface_addr_t *iface_addr;
 
     UCT_EP_PARAMS_CHECK_DEV_IFACE_ADDRS(params);
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
 
+    iface_addr = (const uct_cuda_ipc_iface_addr_t*)params->iface_addr;
+
+    self->keepalive.device = CU_DEVICE_INVALID;
+    uct_cuda_ipc_ep_set_keepalive_device(self);
+
```
Comment (Akshay-Venkatesh) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:53]: Should we throw an error here if no device is found at the time of creating ep?

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 6]
```diff
@@ -165,3 +193,21 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_ipc_ep_put_zcopy,
...
     return status;
 }
+
+ucs_status_t uct_cuda_ipc_ep_check(const uct_ep_h tl_ep, unsigned flags,
+                                   uct_completion_t *comp)
+{
+    uct_cuda_ipc_ep_t *ep = ucs_derived_of(tl_ep, uct_cuda_ipc_ep_t);
+    CUdevice this_device;
+
+    uct_cuda_ipc_ep_set_keepalive_device(ep);
```
Comment (Akshay-Venkatesh) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:203]: @dmitrygx why do we reset ep->keepalive.device here? If current device is invalid or comparison with cached keepalive.device doesn't match, we should throw an error, right?

**Review by Akshay-Venkatesh — COMMENTED** [review]
@dmitrygx just checking but do the tests pass after #5737 was merged?

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 7]
```diff
@@ -22,19 +22,47 @@
 #define UCT_CUDA_IPC_PUT 0
 #define UCT_CUDA_IPC_GET 1
 
+
+static UCS_F_ALWAYS_INLINE void
+uct_cuda_ipc_ep_set_keepalive_device(uct_cuda_ipc_ep_t *ep)
+{
+    if ((ep->keepalive.device == CU_DEVICE_INVALID) &&
+        (UCT_CUDADRV_FUNC(cuCtxGetDevice(&ep->keepalive.device),
+                          UCS_LOG_LEVEL_TRACE) != UCS_OK)) {
```
Comment (Akshay-Venkatesh) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:31]: Instead of cuCtxGetDevice, can we use [cuCtxGetCurrent](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1g8f13165846b73750693640fb3e8380d0) ? 
This we directly compare context pointers and avoid the case of application destroying and creating a new context on the same device. If the original context is destroyed then the peer mappings may have to be purged.

Comment (Akshay-Venkatesh) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:31]: @dmitrygx Added https://github.com/openucx/ucx/pull/6393 to address the above comment. With #6393, calling  UCT_CUDADRV_GET_CTX(_ctx_ret) instead of specifically looking for driver api ctx or primary ctx should serve the above goal.

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 5 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:53]: but then many applications will fail, e.g. if an application wants to use a host memory buffer, but we create CUDA-IPC EP anyway.

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 6 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:203]: the idea is to initialize cached CUDA device in `uct_ep_check()` if it is invalid after `uct_ep_create()`.
and if we can cuCtxGetDevice is now OK, CUDA-IPC caches the value for further checking with a current one.

we need this, because a user may don't want to use CUDA memory at all, but we will have CUDA-IPC lane on which `uct_ep_check()` will be called (but CUDA device won't be initialized in this case)

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 7 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:31]: @Akshay-Venkatesh I see, so the idea is to cache CUDA context instead of CUDA device, right?

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 5 above)
Comment (Akshay-Venkatesh) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:53]: I assumed we create endpoints lazily and defer cuda_ipc ep creation only until a large message device to device transfer is needed.

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 7 above)
Comment (Akshay-Venkatesh) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:31]: Right

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 6 above)
Comment (Akshay-Venkatesh) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:203]: Makes sense is cudaipc ep creation can occur without a device context being present

**Comment by dmitrygx** [comment]
@Akshay-Venkatesh we discussed this PR offline with @yosefe and decided to keep it without checking GPU device for now.
probably, it should be not a part of `uct_ep_check()`, but all UCT EP calls.
wdyt?

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> @Akshay-Venkatesh we discussed this PR offline with @yosefe and decided to keep it without checking GPU device for now.
> probably, it should be not a part of `uct_ep_check()`, but all UCT EP calls.
> wdyt?

If checking GPU device is adding a lot of complexity, then a separate follow up PR is fine by me.

**Comment by dmitrygx** [comment]
> > @Akshay-Venkatesh we discussed this PR offline with @yosefe and decided to keep it without checking GPU device for now.
> > probably, it should be not a part of `uct_ep_check()`, but all UCT EP calls.
> > wdyt?
> 
> If checking GPU device is adding a lot of complexity, then a separate follow up PR is fine by me

I think it's better to discuss it first and then will implement

**Comment by dmitrygx** [comment]
@brminich @Akshay-Venkatesh @bureddy could you review pls?

File: src/uct/base/uct_iface.c [hunk 8]
```diff
@@ -586,3 +586,52 @@ ucs_status_t uct_base_ep_am_short_iov(uct_ep_h ep, uint8_t id, const uct_iov_t *
...
+    ucs_status_t status;
+    ucs_time_t createtime;
+
+    UCT_EP_KEEPALIVE_CHECK_PARAM(flags, comp);
+
+    ucs_assert(ka != NULL);
+
+    status = ucs_sys_get_file_time(ka->proc, UCS_SYS_FILE_TIME_CTIME,
+                                   &createtime);
+    return ucs_likely((status == UCS_OK) && (ka->starttime == createtime)) ?
```
Comment (brminich) [inline src/uct/base/uct_iface.c:634]: minor: maybe rewrite with `if/else`, looks confusing now

File: src/uct/sm/mm/base/mm_ep.c [hunk 9]
```diff
@@ -523,37 +523,16 @@ ucs_status_t uct_mm_ep_flush(uct_ep_h tl_ep, unsigned flags,
...
 
-        ep->keepalive = ucs_malloc(sizeof(*ep->keepalive) + proc_len + 1,
-                                   "mm_ep->keepalive");
+    if (ucs_unlikely(ep->keepalive == NULL)) {
+        ep->keepalive = uct_ep_keepalive_create(ep->fifo_ctl->owner.pid,
+                                                ep->fifo_ctl->owner.starttime);
         if (ep->keepalive == NULL) {
-            status = UCS_ERR_NO_MEMORY;
-            goto err_set_ep_failed;
+            return uct_iface_handle_ep_err(tl_ep->iface, tl_ep,
```
Comment (brminich) [inline src/uct/sm/mm/base/mm_ep.c:532]: do we really need to call error handler with `UCS_ERR_ENDPOINT_TIMEOUT`? This is local memory problem

**Review by Akshay-Venkatesh — APPROVED** [review]

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 10]
```diff
@@ -22,19 +22,30 @@
...
                                                  uct_cuda_ipc_iface_t);
+    const uct_cuda_ipc_iface_addr_t *iface_addr;
 
     UCT_EP_PARAMS_CHECK_DEV_IFACE_ADDRS(params);
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
 
+    iface_addr = (const uct_cuda_ipc_iface_addr_t*)params->iface_addr;
+
+    self->keepalive = uct_ep_keepalive_create(iface_addr->pid,
+                                              iface_addr->starttime);
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c:38]: can we avoid passing starttime in iface address? this increases iface address size?
instead, keepalive_create will read ctime itself

File: src/uct/sm/mm/base/mm_ep.c [hunk 11]
```diff
@@ -523,37 +523,16 @@ ucs_status_t uct_mm_ep_flush(uct_ep_h tl_ep, unsigned flags,
...
-                                                  ep->fifo_ctl->owner.pid);
-        if (proc_len <= 0) {
-            return UCS_ERR_INVALID_PARAM;
-        }
+    uct_mm_ep_t *ep = ucs_derived_of(tl_ep, uct_mm_ep_t);
 
-        ep->keepalive = ucs_malloc(sizeof(*ep->keepalive) + proc_len + 1,
-                                   "mm_ep->keepalive");
+    if (ucs_unlikely(ep->keepalive == NULL)) {
+        ep->keepalive = uct_ep_keepalive_create(ep->fifo_ctl->owner.pid,
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_ep.c:529]: maybe we remote "uct_ep_keepalive_create" function, keep only "uct_ep_keepalive_check". and "uct_ep_keepalive_check(pid)" will create what's needed on demand

File: test/gtest/uct/test_peer_failure.cc [hunk 12]
```diff
@@ -479,11 +553,25 @@ class test_uct_peer_failure_keepalive : public test_uct_peer_failure
...
+
         if (has_mm()) {
             uct_mm_ep_t *ep = ucs_derived_of(tl_ep, uct_mm_ep_t);
-            ASSERT_NE((void*)NULL, ep->keepalive);
-            ep->keepalive->starttime--;
+
+            ka_info = ep->keepalive;
+            ASSERT_TRUE(ka_info != NULL);
+        } else if (has_cuda_ipc()) {
+#if HAVE_CUDA
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc:565]: why needed?

**Review by yosefe — COMMENTED** [review]
as general comment, according to this PR we will do ep_check both on cuda_ipc and sm?
can we avoid doing it on both?

**Comment by dmitrygx** [comment]
> as general comment, according to this PR we will do ep_check both on cuda_ipc and sm?
> can we avoid doing it on both?

yes, we will do it on both.
but if we are going to add GPU device checking in CUDA-IPC then, `uct_ep_check()` of MM and CUDA-IPC will be different.
So, we could not rely on this.
My idea is to add `UCX_LOCAL_PROC_KEEPALIVE_INTERVAL=60s` that will be used by UCT base implementation of keepalive to check peer existence once in 60s, but it will be not real `60s` since a user may be slower (e.g. `UCP_KEEPALIVE_INTERVAL=120s`).

**Comment by yosefe** [comment]
@dmitrygx i don't think we're going to add GPU device checking; we don't handle a case where GPU is removed (??) but the process is still running. 
also, modifying keepalive interval "under the hood" is not a good practice

File: src/uct/sm/mm/base/mm_ep.c (hunk 11 above)
Comment (yosefe) [reply src/uct/sm/mm/base/mm_ep.c:529]: why cuda ipc does not need this?

**Comment by dmitrygx** [comment]
> @dmitrygx i don't think we're going to add GPU device checking; we don't handle a case where GPU is removed (??) but the process is still running.

we don't handle a case where GPU is down, but a process is still running - we don't call user's error handling callback in this case when asynchronous memory transfer fails.

> also, modifying keepalive interval "under the hood" is not a good practice

I don't mean to change, I meant to have UCT's keepalive interval for checking a peer which is running on the same machine once per UCT internal timeout.

**Comment by yosefe** [comment]
> I don't mean to change, I meant to have UCT's keepalive interval for checking a peer which is running on the same machine once per UCT internal timeout.

Right, but if user set KA interval to 60sec, the expectation is to detect failure within 60 sec, not 120..

**Comment by dmitrygx** [comment]
> > I don't mean to change, I meant to have UCT's keepalive interval for checking a peer which is running on the same machine once per UCT internal timeout.
> 
> Right, but if user set KA interval to 60sec, the expectation is to detect failure within 60 sec, not 120..

yes, but anyway we don't give this, because UCX_KEEPALIVE_INTERVAL per worker value, but not per EP as it can be.
so, if some worker has `N` EPs (i.e. peers), KEEPALIVE functionality will report an error about failed peer after `<= (N - 1) * UCX_KEEPALIVE_INTERVAL`, e.g. `99` minutes if we have `100` peers.

**Comment by dmitrygx** [comment]
@yosefe @brminich could you review pls?

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c [hunk 13]
```diff
@@ -68,34 +68,18 @@ ucs_status_t uct_cuda_ipc_iface_get_device_address(uct_iface_t *tl_iface,
...
 {
-    uct_cuda_ipc_iface_t *iface           =
-            ucs_derived_of(tl_iface, uct_cuda_ipc_iface_t);
-    const uct_cuda_ipc_iface_addr_t *addr =
-            (const uct_cuda_ipc_iface_addr_t*)iface_addr;
+    uct_cuda_ipc_iface_t  *iface = ucs_derived_of(tl_iface, uct_cuda_ipc_iface_t);
 
     return ((uct_cuda_ipc_iface_node_guid(&iface->super) ==
-            *((const uint64_t *)dev_addr)) && (getpid() != addr->pid));
+             *((const uint64_t *)dev_addr)) && ((getpid() != *(pid_t *)iface_addr)));
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.c]: beautify?

File: src/uct/sm/mm/base/mm_md.h [hunk 14]
```diff
@@ -61,7 +61,7 @@ typedef struct uct_mm_md {
 
 
 /* Check if available on current machine */
-typedef ucs_status_t (*uct_mm_mapper_query_func_t)();
+typedef ucs_status_t (*uct_mm_mapper_query_func_t)(int *attach_shm_file_p);
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_md.h:64]: pls describe return value in attach_shm_file_p

File: src/uct/sm/mm/base/mm_md.h [hunk 15]
```diff
@@ -88,6 +88,12 @@ typedef ucs_status_t
                                    uct_mm_remote_seg_t *rseg);
 
 
+/* Clean up the remote segment handle created by mem_attach() */
+typedef void
+(*uct_mm_mapper_mem_detach_func_t)(uct_mm_md_t *md,
+                                   const uct_mm_remote_seg_t *rseg);
+
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_md.h:95]: why moved?

**Review by yosefe — COMMENTED** [review]
BTW, we still would need cuda_ipc error handling support to make it be selected, right?

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c (hunk 13 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_ipc/cuda_ipc_iface.c]: this was in previous commits, rolled back changes
if see all files changed - there is no change

Comment (yosefe) [reply src/uct/cuda/cuda_ipc/cuda_ipc_iface.c]: ok

File: src/uct/sm/mm/base/mm_md.h (hunk 15 above)
Comment (dmitrygx) [reply src/uct/sm/mm/base/mm_md.h]: fixed

File: src/uct/sm/mm/base/mm_md.h (hunk 14 above)
Comment (dmitrygx) [reply src/uct/sm/mm/base/mm_md.h:72]: done

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Comment by dmitrygx** [comment]
@brminich @yosefe readthedocs failed here as well.
can we restart it somehow?

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe ok to merge?
